core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -204,7 +202,117 @@ Map<TopicIdPartition, FetchRequest.PartitionData> acquirablePartitions() {
         return topicPartitionData;
     }
 
-    private void releasePartitionLocks(String groupId, Set<TopicIdPartition> topicIdPartitions) {
+    // In case, fetch offset metadata doesn't exist for any topic partition in the list of topic partitions, we do a
+    // replicaManager.readFromLog to populate the offset metadata.
+    private void maybeUpdateFetchOffsetMetadataForTopicPartitions(Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData) {
+        Map<TopicIdPartition, FetchPartitionOffsetData> replicaManagerReadResponseData = null;
+        for (Map.Entry<TopicIdPartition, FetchRequest.PartitionData> entry : topicPartitionData.entrySet()) {
+            TopicIdPartition topicIdPartition = entry.getKey();
+            SharePartition sharePartition = sharePartitionManager.sharePartition(shareFetchData.groupId(), topicIdPartition);
+            if (sharePartition == null) {
+                log.error("Encountered null share partition for groupId={}, topicIdPartition={}. Skipping it.", shareFetchData.groupId(), topicIdPartition);
+                continue;
+            }
+            if (sharePartition.latestFetchOffsetMetadata() == null) {
+                if (replicaManagerReadResponseData == null) {
+                    replicaManagerReadResponseData = readFromLog(topicPartitionData);
```

Comment:
yeah, but in case there are multiple partitions for which offset metadata is missing, I thought that its better to make a single `readFromLog` call instead of multiple `readFromLog` calls. Thus, using all the topic partitions in one call itself. Wdyt?

Reply:
Maybe we can first make a pass to collect all partitions missing offset metadata and then make a single `readFromLog` with those partitions?